Bootstrap 4

prateekkumarweb opened the pull request:
Signed-off-by: Prateek Kumar <prateek@prateekkumar.in>

### Purpose

The following have been updated:
* Navbar
* Grid system (`sm` -> `md`, `md` -> `lg`, etc.)
* Margin to `<h1>` headings
* Hidden/visibility classes (e.g. `hidden` -> `d-none`)
* Use outline buttons from Bootstrap 4
* Hero Carousel widget
* Pagination
* Margins in buttons
* Button sizes
* Labels changed to Badges
* Filter form in publications page

gcushen commented:
Please can someone help review this? Thanks.

prateekkumarweb commented:
A live site with changes from this PR can be found on my site, https://prateekkumar.in

gcushen commented:
Thanks. I assume that the Bootstrap carousel code in the new `widgets/hero_carousel.html` is going to need converting too.

prateekkumarweb commented:
Sure, I will do that soon.

prateekkumarweb commented:
@gcushen Done.

gcushen commented:
Hopefully we can finalise this and merge it soon. I have not fully reviewed it yet, but some (minor) points so far are:

* `pre` tags and their associated styles should not have been removed - this is unrelated to converting the theme to Bootstrap v4
* the `span` class around `back_to_top` needs to be updated from `pull-right` - I believe the new notation is `float-right`?
* the button styles have not been fully updated - e.g. on-hover text color of buttons is not right
* the button on-hover styles for the dark theme (i.e. when `color_theme = "dark"`) are also wrong
* in the dark theme, the search box text should not be black
* with dark theme, the color of tags is not using the right styles and difficult to read - the `.dark .label-*` styles should be applied

prateekkumarweb commented:
> pre tags and their associated styles should not have been removed - this is unrelated to converting the theme to Bootstrap v4. It has also broken some of the original styling of code boxes.

I had added `background-color: inherit;` from `pre` tag in `academic.css` to fix chroma highlighter. But I didn't notice that it broke highlightjs. So, I have restored it.

There was some issue with cite modal after changing to bootstrap 4.
Using `<pre><code class="modal-body tex"></code></pre>` results in
![image](https://user-images.githubusercontent.com/15064671/44969117-37f48a00-af69-11e8-90a3-9e5ca4891b80.png)
Using `<pre class="modal-body tex"></pre>` fixed it.
![image](https://user-images.githubusercontent.com/15064671/44969152-6c684600-af69-11e8-9234-d6509fa4b7e0.png)
I do not know the reason for this.

> the span class around back_to_top needs to be updated from pull-right - I believe the new notation is float-right?

I changed to `float-right`

> the button styles have not been fully updated - e.g. on-hover text color of buttons is not right
the button on-hover styles for the dark theme (i.e. when color_theme = "dark") are also wrong
in the dark theme, the search box text should not be black
with dark theme, the color of tags is not using the right styles and difficult to read - the .dark .label-* styles should be applied

I haven't checked on other themes properly yet. I will check and update soon.

> what is the purpose of 66bb201 - why are we adding a margin to a few specific h1 elements? If we really need to do this, would it not be better defining a new h1 class for them?

With margin:
![screenshot from 2018-09-03 10-40-40](https://user-images.githubusercontent.com/15064671/44968537-2362c280-af66-11e8-964c-1afc751813b3.png)
Without margin:
![screenshot from 2018-09-03 10-43-50](https://user-images.githubusercontent.com/15064671/44968572-48573580-af66-11e8-9f1b-9a50bdae6a24.png)

In bootstrap 3, h1 had default margin which was removed in bootstrap 4.

gcushen commented:
> In bootstrap 3, h1 had default margin which was removed in bootstrap 4.

Would it not be better to add the margin to Academic's existing style definition for headers then, rather than adding an extra class to each instance of a header?

gcushen commented:
@prateekkumarweb thank you very much for all your hard work on this over the last few months!

I've just fixed the remaining issues that I listed in the above review and merged the PR into master :)
